lnbits/extensions/boltcards/views_api.py

```diff
@@ -0,0 +1,202 @@
+    card_id: str = None,
+    wallet: WalletTypeInfo = Depends(require_admin_key),
+):
+    try:
+        if len(bytes.fromhex(data.uid)) != 7:
+            raise HTTPException(
+                detail="Invalid bytes for card uid.", status_code=HTTPStatus.BAD_REQUEST
+            )
+
+        if len(bytes.fromhex(data.k0)) != 16:
+            raise HTTPException(
+                detail="Invalid bytes for k0.", status_code=HTTPStatus.BAD_REQUEST
+            )
+
+        if len(bytes.fromhex(data.k1)) != 16:
+            raise HTTPException(
+                detail="Invalid bytes for k1.", status_code=HTTPStatus.BAD_REQUEST
+            )
+
+        if len(bytes.fromhex(data.k2)) != 16:
+            raise HTTPException(
+                detail="Invalid bytes for k2.", status_code=HTTPStatus.BAD_REQUEST
+            )
+    except:
+        raise HTTPException(
+            detail="Invalid byte data provided.", status_code=HTTPStatus.BAD_REQUEST
+        )
+
+    if card_id:
+        card = await get_card(card_id)
+        if not card:
+            raise HTTPException(
+                detail="Card does not exist.", status_code=HTTPStatus.NOT_FOUND
+            )
+        if card.wallet != wallet.wallet.id:
+            raise HTTPException(
+                detail="Not your card.", status_code=HTTPStatus.FORBIDDEN
+            )
+        card = await update_card(card_id, **data.dict())
+    else:
+        card = await create_card(wallet_id=wallet.wallet.id, data=data)
+    return card.dict()
+
+
+@boltcards_ext.delete("/api/v1/cards/{card_id}")
+async def api_card_delete(card_id, wallet: WalletTypeInfo = Depends(require_admin_key)):
+    card = await get_card(card_id)
+
+    if not card:
+        raise HTTPException(
+            detail="Card does not exist.", status_code=HTTPStatus.NOT_FOUND
+        )
+
+    if card.wallet != wallet.wallet.id:
+        raise HTTPException(detail="Not your card.", status_code=HTTPStatus.FORBIDDEN)
+
+    await delete_card(card_id)
+    raise HTTPException(status_code=HTTPStatus.NO_CONTENT)
+
+
+@boltcards_ext.get("/api/v1/hits")
+async def api_hits(
+    g: WalletTypeInfo = Depends(get_key_type), all_wallets: bool = Query(False)
+):
+    wallet_ids = [g.wallet.id]
+
+    if all_wallets:
+        wallet_ids = (await get_user(g.wallet.user)).wallet_ids
+
+    cards = await get_cards(wallet_ids)
+    cards_ids = []
+    for card in cards:
+        cards_ids.append(card.id)
+
+    return [hit.dict() for hit in await get_hits(cards_ids)]
+
+
+# /boltcards/api/v1/scan?p=00000000000000000000000000000000&c=0000000000000000
+@boltcards_ext.get("/api/v1/scan")
+async def api_scane(p, c, request: Request):
+    # some wallets send everything as lower case, no bueno
+    p = p.upper()
+    c = c.upper()
+    card = None
+    counter = b""
+
+    # since this route is common to all cards I don't know whitch 'meta key' to use
+    # so I try one by one until decrypted uid matches
+    for cand in await get_all_cards():
```

Comment:
This could become an issue with many registered cards. The bolt server seems to use a single decrypt key for its service: https://github.com/boltcard/boltcard/blob/main/lnurlw_request.go#L66.

An alternative would be to change the API signature to something like: `/api/v1/scan/<card-id>` and and add the <card-id> to the static part of the URI to be programmed to the card. 

I checked how the URIs look on cards programmed by the coincorner app. This is what they look like: `lnurlw://coincorner.io/lnurl/withdraw/auth/563f36e7-faf0-4002-a316-012c29a4759a?picc_data=W2N54LKZYLIKWBXK4HQEFX44UG2VBXKK&cmac=WZEDHGN6TVWO32JT

Not sure but my guess is they use the UUID to map the request to the card.
`

Reply:
I should note that adding a static-uid to the URL would be privacy trade-off. It allows the reader of the card to recognize the individual card. With the current approach only the service can recognize the card.

Reply:
Yeah that was what I considering. I like having common link template for all the card. But it could be  an issue for many card. 

I think it could be optional to have card in the link and therefor 'fast' solo link for every card.

Reply:
@iWarpBTC I did like the initial PR that had both API options available for use.

Reply:
One possible concern I'm having here is with a project I'm working on with Praia Bitcoin (Bitcoin Beach Brasil) to bulk create Bolt Cards on LNBits (say, 500 at a time). For that use case, it's great as it is now, as it reduces the steps manual steps required to get a card set up. If we do put the card UID back into the output, then we'd potentially have to enter the card UID into the template in the android app when programming the card.

Reply:
For clarity, I think there are basically 3 different approaches:

1. SUN parameters only - each card with the same key per service deployment
2. SUN parameters only - each card with custom encryption key
3. SUN parameters + static card-id - each card with custom encryption key

Reply:
I actually like option 1 because it still offers link privacy against the reader, easier mass provisioning of cards and fast lookup. At the cost of server key leakage. But if that can leak, individual card keys can also ...

Reply:
@leesalminen The initial PR had unencrypted option, which is not compatible with bolt-nfc-android-app writter. It would be number 4 approach when I follow up on @titusz's list.

I came to the conclusion that a shared encryption key is not much suitable for a service like lnbits. But we can think about how it could be appropriately implemented for such a service.

On the other hand adding a route for the approach 3 would be quick and painless (if I'm not missing anything).

Reply:
Hey @iWarpBTC, wondering if it'd be possible to put an optional checkbox in the android programming app "append card UID to url" in the "Write NFC" screen there? Then we could offer 2 routes in LNBits: `/api/v1/scan` (as it is now) and `/api/v1/scan/{card_uid}`.

Reply:
I think it would be possible but we don't need that. We can write there the card's UID or even our own ID manually now. We can have anything before `?p=00000000000000000000000000000000&c=0000000000000000`. I hope the link (with or without card's ID) could be part of the QR in the future. Before that, we can display the link that should be written in the card's details in lnbits (maybe with a copy button).

Reply:
I just pushed up to the branch so that it will allow both ways `/boltcards/api/v1/scan` where it loops over all the cards and decrypts (best privacy, slow if lots of cards in db) and `/boltcards/api/v1/scan/{card_uid}` where it only fetches and decrypts with 1 card record (small potential privacy leak, more efficient)

Reply:
I'm just wondering if revealing the UID and guessing the counter could endanger the encryption key... Unfortunately I'm not that familiar with AES...

Reply:
> I'm just wondering if revealing the UID and guessing the counter could endanger the encryption key... Unfortunately I'm not that familiar with AES... But probably I'm being too paranoid here.

I'm definitely not qualified to make anything approaching an authoritative statement on that.

In an abundance of caution, we could change this to use LNBit's internal card ID (not used in the encryption) instead.

The only downside I see to that is that when a user programs the cards, they'll need to use this opaque identifier only found in LNBits. A future update to the Android app though as you mention could resolve that UX concern by embedding it into the QR code they scan.

Reply:
Pushed this

Reply:
Crap, I forgot that breez wallet lowercases everything in the string, so lookup on the card_id fails. I'm back to card_uid?

Reply:
Ouch! I did not realize card_id is case sens. So it is card_uid or a new dedicated id. Up to you I guess.

Reply:
> Ouch! I did not realize card_id is case sens. So it is card_uid or a new dedicated id. Up to you I guess.

I put the card_uid back in place.